File: 103_Binary_Tree_Zigzag_Level_Order_Traversal/103_1.java
import java.util.List;
import java.util.LinkedList;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int v) { val = v; }
}

class Solution {
	public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        if (root == null) return new LinkedList<List<Integer>>();
        List<LinkedList<Integer>> res = new LinkedList<>();
        preOrder(root, 0, res);
        return (List)res;
    }

    void preOrder(TreeNode node, int depth, List<LinkedList<Integer>> res) {
        if (depth >= res.size()) res.add(new LinkedList<Integer>());
        if (depth % 2 == 0)     res.get(depth).offer(node.val);
        else                    res.get(depth).offerFirst(node.val);
        if (node.left  != null) preOrder(node.left,  depth + 1, res);
        if (node.right != null) preOrder(node.right, depth + 1, res);
    }
}

public class leetcode {
    public static void main(String[] args) {
        TreeNode node1 = new TreeNode(1);
        TreeNode node2 = new TreeNode(2);
        TreeNode node3 = new TreeNode(3);
        TreeNode node4 = new TreeNode(4);
        TreeNode node5 = new TreeNode(5);
        node1.left  = node2;
        node1.right = node3;
        node2.left  = node4;
        node3.right = node5;
        Solution s = new Solution();
        List<List<Integer>> res = s.zigzagLevelOrder(node1);
        for (List<Integer> ll : res) {
            for (Integer l : ll)
                System.out.print(l + " ");
            System.out.println();
        }
    }
}

File: 103_Binary_Tree_Zigzag_Level_Order_Traversal/103_1.cpp
#include <iostream>
#include <vector>
#include <deque>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x): val(x), left(NULL), right(NULL) { }
};

class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if (!root) return vector<vector<int>>();
        vector<vector<int>> res;
        deque<TreeNode*> queue;
        queue.emplace_back(root);
        queue.emplace_back(nullptr);
        deque<int> resIn;
        bool is_left = true;
        while (!queue.empty()) {
            TreeNode *node = queue.front();
            queue.pop_front();
            if (node) {
                if (is_left)     resIn.emplace_back(node->val);
                else             resIn.emplace_front(node->val);
                if (node->left)  queue.emplace_back(node->left);
                if (node->right) queue.emplace_back(node->right);
            } else {
                res.emplace_back(begin(resIn), end(resIn));
                resIn.clear();
                if (!queue.empty()) queue.emplace_back(nullptr);
                is_left = !is_left;
            }
        }
        return res;
    }
};


int main() {
    TreeNode *node1 = new TreeNode(1);
    TreeNode *node2 = new TreeNode(2);
    TreeNode *node3 = new TreeNode(3);
    TreeNode *node4 = new TreeNode(4);
    TreeNode *node5 = new TreeNode(5);
    node1->left  = node2;
    node1->right = node3;
    node2->left  = node4;
    node3->right = node5;
    Solution s;
    for (const auto &i : s.zigzagLevelOrder(node1)) {
        for (const auto &j : i) {
            cout << j << " ";
        }
        cout << "\n";
    }
}